### DISCUSSION TITLE
increasing ti.max_tries doesn't add more retries

### DISCUSSION THREAD

**Discussion by karunpoudel-chr** [apache/airflow#24997] [General] [answered]
### Apache Airflow version

2.2.5

### What happened

A task (PythonOperator) had `'retries': 1`. When the task runs, depending upon conditions, it dynamically increases it's max_tries by using `kwargs['ti'].max_tries += 1` and the task fails. Airflow only retried once based on what was set in the Operator. It did retry more than once as it should have based on new ti.max_tries.
This was working in 2.1.4 but stopped working after 2.2.0.

### What you think should happen instead

Airflow should have retried the task 2nd time also since the new ti.max_tries is 2.
This was working in 2.1.4 but stopped working after 2.2.0.

### How to reproduce

The following dag retries only once, whereas in 2.1.4 it retires 4 more times.

```
from datetime import timedelta
import pendulum
from airflow import DAG
from airflow.operators.python import PythonOperator

default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'email_on_failure': True,
    'email_on_retry': False,
    'retries': 1,
    'retry_delay': timedelta(seconds=10),
}


def main(**kwargs):
    print("hello")
    try_number = kwargs['ti'].try_number
    print(f"Try = {try_number}")
    some_condition = True
    if some_condition and try_number <= 4:
        print(f"Max = {kwargs['ti'].max_tries}")
        print('Increasing max_tries by 1')
        kwargs['ti'].max_tries += 1
        raise Exception("Retry again")


with DAG('test_retry',
         start_date=pendulum.datetime(2021, 1, 1, tz='America/Chicago'),
         max_active_runs=1,
         schedule_interval=None,
         default_args=default_args,
         catchup=False
         ) as dag:

    test_retry= PythonOperator(
        task_id="test_retry",
        python_callable=main,
        provide_context=True
    )
```

### Operating System

Debian GNU/Linux 11 (bullseye)

### Deployment

Astronomer

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Even if it worked before, that was purely accidental. This is almost certainly not an intention for the kwargs of tasks to be modified dynamically and it was never supposed to work like that.

**Comment by potiuk** [comment]
Converted it into discussion in case you would like to discuss on various ways similar things could be achieved, but it's for sure not going to work this way in the future,

**Comment by karunpoudel-chr** [comment]
Anyone has idea on how to achieve this?
Basically, normally we just want 1 retries (no point in multiple retry if it is stuck in same place). But there is 1 particular error scenario when we want it to retry multiple times (lets say, max retry of 5 in this scenario) as long that particular error is reoccurring.
And don't want alert email as long as it is retrying.

**Reply by potiuk** [reply, marked as answer]
Set maximum number of retries and if you do not want to retry any more, throw AirlfowFailException. It's been especially designed to stop retry loop:

```
class AirflowFailException(AirflowException):
    """Raise when the task should be failed without retrying."""
```
